File: Cpp/Server/SimpleServer.cpp
#include <iostream>
#include <boost/asio.hpp>

using boost::asio::ip::tcp;

int main() {
    try {
        boost::asio::io_context io_context;

        // Create and bind the acceptor to the desired address and port
        tcp::acceptor acceptor(io_context, tcp::endpoint(tcp::v4(), 8080));

        std::cout << "Server running on 127.0.0.1:8080" << std::endl;

        while (true) {
            // Accept incoming connections
            tcp::socket socket(io_context);
            acceptor.accept(socket);

            // Read the length of the incoming message
            uint32_t message_length;
            boost::asio::read(socket, boost::asio::buffer(&message_length, sizeof(message_length)));

            // Read the body of the message
            std::string message_body(message_length, '\0');
            boost::asio::read(socket, boost::asio::buffer(&message_body[0], message_length));

            // Echo the message back to the client
            boost::asio::write(socket, boost::asio::buffer(&message_length, sizeof(message_length)));
            boost::asio::write(socket, boost::asio::buffer(message_body));
        }
    } catch (std::exception& e) {
        std::cerr << e.what() << std::endl;
    }

    return 0;
}

File: Rust/Server/simple_server.rs
use tokio::net::{TcpListener, TcpStream};
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let listener = TcpListener::bind("127.0.0.1:8080").await?;
    println!("Server running on 127.0.0.1:8080");

    loop {
        let (mut socket, _) = listener.accept().await?;

        tokio::spawn(async move {
            loop {
                // Read the length of the incoming message
                let mut len_buf = [0; 4];
                if let Err(_) = socket.read_exact(&mut len_buf).await {
                    return;
                }
                let len = u32::from_le_bytes(len_buf);

                // Read the body of the message
                let mut buf = vec![0; len as usize];
                if let Err(_) = socket.read_exact(&mut buf).await {
                    return;
                }

                // Echo the message back to the client
                if let Err(_) = socket.write_all(&len_buf).await {
                    return;
                }
                if let Err(_) = socket.write_all(&buf).await {
                    return;
                }
            }
        });
    }
}
